subprojects/plugins/src/integTest/groovy/org/gradle/api/tasks/JavaExecSystemPropertiesIntegrationTest.groovy

```diff
@@ -0,0 +1,179 @@
+    def setup() {
+        file("src/main/java/org/gradle/demo/Greeter.java") << """package org.gradle.demo;
+
+    public class Greeter {
+        public static String getGreeting() {
+            return "Hello, " + System.getProperty("name", "unidentified person") + "!";
+        }
+        public static void main(String... args) {
+            System.out.println(getGreeting());
+        }
+    }
+"""
+        file("src/test/java/org/gradle/demo/GreeterTest.java") << """
+package org.gradle.demo;
+
+    public class GreeterTest {
+        @org.junit.Test public void testGreeting() {
+            org.junit.Assert.assertEquals("Hello, Aditi!", new Greeter().getGreeting());
+        }
+    }
+"""
+        buildFile << """
+            apply plugin: 'application'
+
+            application {
+                mainClass = 'org.gradle.demo.Greeter'
+            }
+
+            repositories {
+                 ${mavenCentralRepository()}
+            }
+
+            dependencies {
+                 testImplementation 'junit:junit:4.13'
+            }
+
+        """
+    }
+
+    def "can use a provider as a system property value"() {
+        buildFile << """
+            tasks.named('run') {
+                systemProperty('name', providers.provider { "Aditi" })
+            }
+        """
+
+        when:
+        run "run"
+
+        then:
+        outputContains ("Hello, Aditi!")
+    }
+
+    def "up-to-date checking considers providers in system properties"() {
+        buildFile << """
+            tasks.named('test', Test) {
+                systemProperty('name', providers.systemProperty('name'))
+            }
+        """
+
+        when:
+        run "test", "-Dname=Aditi"
+
+        then:
+        executedAndNotSkipped (":test")
+
+        when:
+        run "test", "-Dname=Aditi"
+
+        then:
+        skipped (":test")
+
+        when:
+        fails "test", "-Dname=Wajeeha"
+
+        then:
+        executedAndNotSkipped (":test")
+        outputContains "org.gradle.demo.GreeterTest > testGreeting FAILED"
+
+    }
+
+    def "tasks inputs are tracked when using providers as system property values"() {
+        buildFile << """
+            def buildName = tasks.register("buildName", BuildName) {
+                outputFile = layout.buildDirectory.file("name.txt")
+                person = providers.systemProperty("name").orElse("unidentified person")
+            }
+
+            abstract class BuildName extends DefaultTask {
+                @Input
+                abstract Property<String> getPerson()
+
+                @OutputFile
+                abstract RegularFileProperty getOutputFile()
+
+                @TaskAction
+                void execute() {
+                    outputFile.asFile.get().text = person.get()
+                }
+            }
+
+            tasks.named('run') {
+                systemProperty('name', buildName.map { it.outputFile.asFile.get().text })
```

Comment:
Keeping the provider chain intact fixes it:
```groovy
systemProperty('name', buildName.flatMap { it.outputFile }.map { it.asFile.text })
 ```

Reply:
Is it expected? I thought that using `buildName.map { .. }` would be enough to preserve the task dependency? Why isn't it executed?

Reply:
Ok, this is **not** expected, relevant issue: https://github.com/gradle/gradle/issues/19252